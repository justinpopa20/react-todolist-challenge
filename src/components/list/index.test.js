import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import List from './'

describe('List', () => {
  const props = {
    items: [],
    fetchAllListItems: jest.fn(),
    editItem: jest.fn(),
    deleteItem: jest.fn()
  }

  props.fetchAllListItems.mockReturnValue([])

  const mockItems = [{ id: 0, text: 'Todo 1'}, { id: 1, text: 'Todo 2'}]

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render items', () => {
    const wrapper = render(<List {...{...props, items: mockItems}} />)

    expect(wrapper.baseElement).toMatchSnapshot()
    expect(props.fetchAllListItems).toHaveBeenCalled()
    expect(wrapper.getAllByTestId('todo-item').length).toBe(mockItems.length)
  })

  test('should display loading state', () => {
    const wrapper = render(<List isFetching={true} {...props} />)

    expect(wrapper.getAllByText('Loading...').length).toBe(1)
  })

  // test('should display error state', () => {
  //   const wrapper = render(<List isFetching={false} error="Error text" {...props} />)

  //   expect(wrapper.getAllByText('Error text').length).toBe(1)
  // })

  test('should edit todo', () => {
    const wrapper = render(<List isFetching={false} {...{...props, items: mockItems}} />)

    const editBtn = wrapper.getAllByTestId("edit-btn")[1]
    
    fireEvent.click(editBtn)

    const input = wrapper.getByTestId("edit-input")
    fireEvent.change(input, { target: { value: 'test123' } })

    fireEvent.click(editBtn)

    expect(props.editItem).toHaveBeenCalledWith(mockItems[1], { text: "test123"})

  })

  test('should cancel edit todo', () => {
    const wrapper = render(<List isFetching={false} {...{...props, items: mockItems}} />)

    const editBtn = wrapper.getAllByTestId("edit-btn")[1]

    fireEvent.click(editBtn)
    const cancelBtn = wrapper.getAllByTestId("delete-btn")[1]
    fireEvent.click(cancelBtn)
    fireEvent.click(editBtn)

    expect(props.editItem).not.toHaveBeenCalled()
  })

  test('should delete todo', () => {
    const wrapper = render(<List isFetching={false} {...{...props, items: mockItems}} />)

    const deleteBtn = wrapper.getAllByTestId("delete-btn")[1]
    fireEvent.click(deleteBtn)

    expect(props.deleteItem).toHaveBeenCalledWith(mockItems[1], 1)
  })
})