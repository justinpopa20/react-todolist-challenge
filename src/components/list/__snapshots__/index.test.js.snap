// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render items 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            width="66%"
          >
            Item
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div
              class="todo-item"
              data-testid="todo-item"
            >
              <div
                class="todo-item-content"
              >
                Todo 1
              </div>
              <div
                class="todo-item-status"
              />
            </div>
          </td>
          <td>
            <button
              class="button edit"
              data-testid="edit-btn"
            >
              Edit
            </button>
            <button
              class="button delete"
              data-testid="delete-btn"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <div
              class="todo-item"
              data-testid="todo-item"
            >
              <div
                class="todo-item-content"
              >
                Todo 2
              </div>
              <div
                class="todo-item-status"
              />
            </div>
          </td>
          <td>
            <button
              class="button edit"
              data-testid="edit-btn"
            >
              Edit
            </button>
            <button
              class="button delete"
              data-testid="delete-btn"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
